File: lib/resolution.ml
(** resolution.ml — ENGINE 3: Resolution Refutation Prover.
    
    Implements a first-order resolution prover using the Otter loop
    (given-clause algorithm) with binary resolution, factoring, and
    subsumption.
    
    References:
    - Robinson (1965), "A Machine-Oriented Logic Based on the Resolution Principle".
    - McCune (2003), "Otter 3.3 Reference Manual".
    - Harrison (2009), "Handbook of Practical Logic", Chapter 4. *)

open Types
open Formula
open Unification

(* ================================================================ *)
(* UTILITIES                                                         *)
(* ================================================================ *)

(** Rename variables in a clause to avoid collisions during resolution.
    Appends a suffix like "_1" or "_2". *)
let rename_clause suffix clause =
  let rename_term = let rec aux = function
    | Var x -> Var (x ^ suffix)
    | Func (f, args) -> Func (f, List.map aux args)
    in aux
  in
  let rename_lit = function
    | Pos (p, args) -> Pos (p, List.map rename_term args)
    | Neg (p, args) -> Neg (p, List.map rename_term args)
  in
  List.map rename_lit clause

(** Check if a clause is a tautology (contains P and ¬P). *)
let is_tautology clause =
  List.exists (fun l1 ->
    match l1 with
    | Pos (p, args1) ->
      List.exists (function Neg (q, args2) -> p = q && args1 = args2 | _ -> false) clause
    | Neg (p, args1) ->
      List.exists (function Pos (q, args2) -> p = q && args1 = args2 | _ -> false) clause
  ) clause

(** Remove duplicate literals from a clause. *)
let simplify_clause clause =
  unique clause

(* ================================================================ *)
(* SUBSUMPTION                                                       *)
(* ================================================================ *)

(** Check if clause c1 subsumes c2.
    c1 subsumes c2 if there exists a substitution σ such that σ(c1) ⊆ c2. *)
let subsumes c1 c2 =
  let rec search sigma = function
    | [] -> true
    | l1 :: rest ->
      (* Try to find a literal in c2 that l1 can unify with *)
      let rec try_l2s = function
        | [] -> false
        | l2 :: l2s ->
          (match unify_literals sigma l1 l2 with
           | Some sigma' -> if search sigma' rest then true else try_l2s l2s
           | None -> try_l2s l2s)
      in
      try_l2s c2
  in
  if List.length c1 > List.length c2 then false
  else search [] c1

(* ================================================================ *)
(* RESOLUTION & FACTORING                                            *)
(* = *)

(** Perform binary resolution on two clauses.
    Returns a list of all possible resolvents. *)
let resolve c1 c2 =
  let c1' = rename_clause "_L" c1 in
  let c2' = rename_clause "_R" c2 in
  let res = ref [] in
  List.iter (fun l1 ->
    List.iter (fun l2 ->
      match unify_complementary [] l1 l2 with
      | Some sigma ->
        let r = List.filter (fun l -> l <> l1) c1' @ List.filter (fun l -> l <> l2) c2' in
        res := (simplify_clause (apply_clause sigma r)) :: !res
      | None -> ()
    ) c2'
  ) c1';
  !res

(** Perform factoring on a clause.
    Returns a list of all possible factors. *)
let factor clause =
  let res = ref [] in
  let rec aux = function
    | [] -> ()
    | l1 :: rest ->
      List.iter (fun l2 ->
        if l1 <> l2 then
          match unify_literals [] l1 l2 with
          | Some sigma ->
            res := (simplify_clause (apply_clause sigma clause)) :: !res
          | None -> ()
      ) rest;
      aux rest
  in
  aux clause;
  !res

(* ================================================================ *)
(* OTTER LOOP (GIVEN-CLAUSE ALGORITHM)                               *)
(* ================================================================ *)

let prove formula timeout_ms =
  let start_time = Unix.gettimeofday () in
  let stats = make_stats () in
  
  (* 1. Preprocessing: Negate -> CNF *)
  let clauses = Cnf.to_clauses (Not formula) in
  let unprocessed = ref (List.filter (fun c -> not (is_tautology c)) clauses) in
  let processed = ref [] in
  
  stats.clauses_gen <- List.length !unprocessed;

  let rec loop () =
    let elapsed = (Unix.gettimeofday () -. start_time) *. 1000. in
    if elapsed > timeout_ms then
      { engine = Resolution; solved = false; stats = { stats with time_ms = elapsed } }
    else if !unprocessed = [] then
      { engine = Resolution; solved = false; stats = { stats with time_ms = elapsed } }
    else
      (* Select "given clause": pick the smallest for heuristic *)
      let given = 
        let sorted = List.sort (fun c1 c2 -> compare (List.length c1) (List.length c2)) !unprocessed in
        List.hd sorted
      in
      unprocessed := List.filter (fun c -> c != given) !unprocessed;
      
      if given = [] then (
        stats.time_ms <- (Unix.gettimeofday () -. start_time) *. 1000.;
        { engine = Resolution; solved = true; stats }
      ) else if List.exists (fun p -> subsumes p given) !processed then
        loop () (* Forward subsumption *)
      else (
        stats.steps <- stats.steps + 1;
        
        (* Backward subsumption: remove clauses in processed subsumed by given *)
        let old_processed_count = List.length !processed in
        processed := List.filter (fun p -> not (subsumes given p)) !processed;
        stats.subsumptions <- stats.subsumptions + (old_processed_count - List.length !processed);
        
        (* Generate new resolvents *)
        let new_resolvents = List.concat_map (resolve given) !processed in
        (* Generate factors *)
        let new_factors = factor given in
        let newborns = new_resolvents @ new_factors in
        
        stats.clauses_gen <- stats.clauses_gen + List.length newborns;
        
        let any_empty = ref false in
        List.iter (fun r ->
          if r = [] then any_empty := true
          else if not (is_tautology r) && 
                  not (List.exists (fun p -> subsumes p r) !processed) &&
                  not (List.exists (fun u -> subsumes u r) !unprocessed) then
            unprocessed := r :: !unprocessed
        ) newborns;
        
        processed := given :: !processed;
        
        if !any_empty then (
          stats.time_ms <- (Unix.gettimeofday () -. start_time) *. 1000.;
          { engine = Resolution; solved = true; stats }
        ) else loop ()
      )
  in
  loop ()

File: python/resolution.py
"""
resolution.py — ENGINE 3: Resolution Refutation Prover.
Ported from resolution.ml.
"""

import time
from typing import List, Tuple, Optional, Set
from fol_types import Term, Variable, Function, Constant, Atom, Negation, Formula
from unification import apply, unify
from formula_utils import unique, free_vars
import cnf

# ================================================================
# TYPES
# ================================================================

class Literal:
    def __init__(self, formula: Formula):
        self.formula = formula
        if isinstance(formula, Atom):
            self.positive = True
            self.predicate = formula.predicate
            self.args = formula.args
        elif isinstance(formula, Negation) and isinstance(formula.formula, Atom):
            self.positive = False
            self.predicate = formula.formula.predicate
            self.args = formula.formula.args
        else:
            raise ValueError(f"Invalid literal formula: {formula}")

    def negate(self) -> 'Literal':
        if self.positive:
            return Literal(Negation(self.formula))
        else:
            return Literal(self.formula.formula)

    def apply_sigma(self, sigma: dict) -> 'Literal':
        def subst_term(t: Term) -> Term:
            if isinstance(t, Variable):
                if t.name in sigma: return subst_term(sigma[t.name])
                return t
            if isinstance(t, Function):
                return Function(t.name, tuple(subst_term(arg) for arg in t.args))
            return t
        
        new_args = tuple(subst_term(arg) for arg in self.args)
        if self.positive:
            return Literal(Atom(self.predicate, new_args))
        else:
            return Literal(Negation(Atom(self.predicate, new_args)))

    def __eq__(self, other):
        if not isinstance(other, Literal): return False
        return self.positive == other.positive and self.predicate == other.predicate and self.args == other.args

    def __hash__(self):
        return hash((self.positive, self.predicate, self.args))

    def __repr__(self):
        return f"{'' if self.positive else '~'}{self.predicate}({', '.join(map(str, self.args))})"

Clause = List[Literal]

# ================================================================
# UTILITIES
# ================================================================

def rename_clause(suffix: str, clause: Clause) -> Clause:
    def rename_term(t: Term) -> Term:
        if isinstance(t, Variable): return Variable(t.name + suffix)
        if isinstance(t, Function): return Function(t.name, tuple(rename_term(arg) for arg in t.args))
        return t
    
    new_clause = []
    for lit in clause:
        new_args = tuple(rename_term(arg) for arg in lit.args)
        if lit.positive:
            new_clause.append(Literal(Atom(lit.predicate, new_args)))
        else:
            new_clause.append(Literal(Negation(Atom(lit.predicate, new_args))))
    return new_clause

def is_tautology(clause: Clause) -> bool:
    positives = { (l.predicate, l.args) for l in clause if l.positive }
    negatives = { (l.predicate, l.args) for l in clause if not l.positive }
    return bool(positives & negatives)

def simplify_clause(clause: Clause) -> Clause:
    return unique(clause)

# ================================================================
# SUBSUMPTION
# ================================================================

def subsumes(c1: Clause, c2: Clause) -> bool:
    def search(sigma: dict, remaining_c1: Clause) -> bool:
        if not remaining_c1: return True
        l1 = remaining_c1[0]
        rest = remaining_c1[1:]
        
        for l2 in c2:
            if l1.positive == l2.positive and l1.predicate == l2.predicate:
                sigma_prime = unify(sigma, list(zip(l1.args, l2.args)))
                if sigma_prime is not None:
                    if search(sigma_prime, rest): return True
        return False

    if len(c1) > len(c2): return False
    return search({}, c1)

# ================================================================
# RESOLUTION & FACTORING
# ================================================================

def resolve(c1: Clause, c2: Clause) -> List[Clause]:
    c1_prime = rename_clause("_L", c1)
    c2_prime = rename_clause("_R", c2)
    resolvents = []
    
    for l1 in c1_prime:
        for l2 in c2_prime:
            if l1.positive != l2.positive and l1.predicate == l2.predicate:
                sigma = unify({}, list(zip(l1.args, l2.args)))
                if sigma is not None:
                    r = [l for l in c1_prime if l != l1] + [l for l in c2_prime if l != l2]
                    r_applied = [l.apply_sigma(sigma) for l in r]
                    resolvents.append(simplify_clause(r_applied))
    return resolvents

def factor(clause: Clause) -> List[Clause]:
    factors = []
    for i, l1 in enumerate(clause):
        for l2 in clause[i+1:]:
            if l1.positive == l2.positive and l1.predicate == l2.predicate:
                sigma = unify({}, list(zip(l1.args, l2.args)))
                if sigma is not None:
                    f_applied = [l.apply_sigma(sigma) for l in clause]
                    factors.append(simplify_clause(f_applied))
    return factors

# ================================================================
# OTTER LOOP
# ================================================================

class ResolutionProver:
    def __init__(self, timeout_ms: float = 50000.0):
        self.timeout_ms = timeout_ms

    def prove(self, formula: Formula) -> Tuple[bool, dict]:
        start_time = time.time()
        steps = 0
        clauses_gen = 0
        subsumptions = 0
        
        # Preprocessing: Negate -> CNF
        negated = Negation(formula)
        raw_clauses = cnf.to_clauses(negated)
        
        unprocessed = []
        for rc in raw_clauses:
            cl = [Literal(f) for f in rc]
            if not is_tautology(cl):
                unprocessed.append(cl)
        
        processed = []
        clauses_gen = len(unprocessed)

        while unprocessed:
            elapsed = (time.time() - start_time) * 1000.0
            if elapsed > self.timeout_ms:
                return False, {"steps": steps, "clauses_gen": clauses_gen, "subsumptions": subsumptions, "time_ms": elapsed}
            
            # Select "given clause" (heuristic: smallest)
            unprocessed.sort(key=len)
            given = unprocessed.pop(0)
            
            if not given:
                elapsed = (time.time() - start_time) * 1000.0
                return True, {"steps": steps, "clauses_gen": clauses_gen, "subsumptions": subsumptions, "time_ms": elapsed}
            
            # Forward subsumption
            if any(subsumes(p, given) for p in processed):
                continue
            
            steps += 1
            
            # Backward subsumption
            old_processed_count = len(processed)
            processed = [p for p in processed if not subsumes(given, p)]
            subsumptions += (old_processed_count - len(processed))
            
            # Generate new resolvents
            newborns = []
            for p in processed:
                newborns.extend(resolve(given, p))
            
            # Generate factors
            newborns.extend(factor(given))
            
            clauses_gen += len(newborns)
            
            for r in newborns:
                if not r:
                    elapsed = (time.time() - start_time) * 1000.0
                    return True, {"steps": steps, "clauses_gen": clauses_gen, "subsumptions": subsumptions, "time_ms": elapsed}
                
                if not is_tautology(r) and \
                   not any(subsumes(p, r) for p in processed) and \
                   not any(subsumes(u, r) for u in unprocessed):
                    unprocessed.append(r)
            
            processed.append(given)

        elapsed = (time.time() - start_time) * 1000.0
        return False, {"steps": steps, "clauses_gen": clauses_gen, "subsumptions": subsumptions, "time_ms": elapsed}
